keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlResourceRepository.kt

```diff
@@ -183,90 +191,83 @@ open class SqlResourceRepository(
     }
   }
 
-  override fun eventHistory(id: String, limit: Int): List<ResourceEvent> {
+  override fun eventHistory(id: String, limit: Int): List<ResourceHistoryEvent> {
     require(limit > 0) { "limit must be a positive integer" }
+
+    val resource = get(id)
+
     return sqlRetry.withRetry(READ) {
       jooq
         .select(EVENT.JSON)
-        .from(EVENT, RESOURCE)
-        .where(EVENT.SCOPE.eq(Scope.RESOURCE.name))
-        .and(RESOURCE.ID.eq(id))
-        .and(RESOURCE.UID.eq(EVENT.REF))
+        .from(EVENT)
+        // look for resource events that match the resource...
+        .where(EVENT.SCOPE.eq(Scope.RESOURCE.name)
+          .and(EVENT.REF.eq(resource.uid))
+        )
+        // ...or application events that match the application as they apply to all resources
+        .or(EVENT.SCOPE.eq(Scope.APPLICATION.name)
+          .and(EVENT_JSON_APPLICATION.eq(resource.application))
+        )
```

Comment:
Would it make sense to add an `application` column to the table to make this easier?

Reply:
Probably, yes.

Reply:
BTW, do you know how much of a performance hit we get from querying within JSON blobs? I know MySQL has optimizations for it, but that's the extend of my knowledge...

Reply:
No idea. I suspect that's something we need to measure and optimize if/when we feel like we need to

Reply:
Added https://github.com/spinnaker/keel/issues/1171 to track.